### ISSUE TITLE
snmpd.conf's rouser (and rwuser) directive does not work if its USER contains a double quote (escaped by \)

### ISSUE THREAD

**Issue by tzhang06** [net-snmp/net-snmp#411]
For instance, with the following rouser directive, snmpd reads the user name fine.:
`  rouser "m#a\"z'u m$a\"z'u" authpriv`

However, when adding a group directive, it doesn't escape the embedded double quote:
`  group grpm_a_z_u_m_a_z_u usm "m#a"z'u m$a"z'u"`

This causes the authorization error of the USER. For instance, from snmpget:
```
  Error in packet
  Reason: authorizationError (access denied to that object)
```
because the USER is not a member of the group.

It needs to be:
`  group grpm_a_z_u_m_a_z_u usm "m#a\"z'u m$a\"z'u"`
(or HEX, 0x6d2361227a2775206d2461227a2775?)

The issue is at vacm_create_simple() in agent/mibgroup/mibII/vacm_conf.c.

Here is the debug trace. 

```
read_config(): read_config.c, 848:
9:read_config:line:
/etc/snmp/snmpd.conf:98 examining: rouser "m#a\"z'u m$a\"z'u" authpriv
trace:
read_config(): read_config.c, 976:
read_config:line:
/etc/snmp/snmpd.conf:98 examining: rouser "m#a\"z'u m$a\"z'u" authpriv
trace:
run_config_handler(): read_config.c, 542:
read_config:parser:
Found a parser.  Calling it: rouser / "m#a\"z'u m$a\"z'u" authpriv
trace:
vacm_create_simple(): mibgroup/mibII/vacm_conf.c, 871:
rouser:
setting auth level: "authpriv"
trace:
vacm_create_simple(): mibgroup/mibII/vacm_conf.c, 1013:
rouser:
passing: group grpm_a_z_u_m_a_z_u usm "m#a"z'u m$a"z'u"
trace:
vacm_create_simple(): mibgroup/mibII/vacm_conf.c, 1052:
rouser:
passing: access grpm_a_z_u_m_a_z_u "" usm authpriv prefix _all_ none none
trace:
read_config(): read_config.c, 848:
9:read_config:line:
```

As a workaround, the rouser directive needs to be replaced by, something like:
```
  group remoteROGroup usm "m#a\"z'u m$a\"z'u"
  access remoteROGroup "" usm authpriv prefix _all_ none none
```

**Comment by fenner** [comment]
You're really talking about whatever created the configuration file for you, right?  The rouser directive works fine if you quote the arguments properly, whether using the hex encoding or something else?

**Comment by tzhang06** [comment]
The group and access directives are automatically added (behind the scene) when you use the rouser (convenient) directive. In the aforementioned example, the properly escaped USER directive has been correctly read from snmpd.conf but the code failed to create the proper group and access directives in the running config of the SNMP daemon.

In vacm_create_simple():

            strlcpy(secname, community, sizeof(secname));
            /*
             * sec->group mapping
             */
            /*
             * group   anonymousGroupNameNUM  any      anonymousSecNameNUM
             */
            snprintf(grpname, sizeof(grpname), "grp%.28s", secname);
            for (tmp=grpname; *tmp; tmp++)
                if (!isalnum((unsigned char)(*tmp)))
                    *tmp = '_';
            snprintf(line, sizeof(line),
                     "%s %s \"%s\"", grpname, model, secname);
            line[ sizeof(line)-1 ] = 0;
            DEBUGMSGTL((token, "passing: %s %s\n", "group", line));
            vacm_parse_group("group", line);
 
...

        /* Use the simple line access command */
        /*
         * access  anonymousGroupNameNUM  "" MODEL AUTHTYPE prefix anonymousViewNUM [none/anonymousViewNUM] [none/anonymousViewNUM]
         */
        snprintf(line, sizeof(line),
                 "%s %s %s %s %s %s %s %s",
                 grpname, context[0] ? context : "\"\"",
                 model, authlevel,
                (ctxprefix ? "prefix" : "exact"),
                 viewname, rw, rw);
        line[ sizeof(line)-1 ] = 0;
        DEBUGMSGTL((token, "passing: %s %s\n", "access", line));
        vacm_parse_access("access", line);

**Comment by tzhang06** [comment]
The use of read_config_save_octet_string() is fine as long as read_config_read_octet_string() is also used by vacm_parse_group(). (copy_nword() doesn’t handle the hex string.)

**Comment by tzhang06** [comment]
BTW, there is another bug here. The auto-generated group name is “grp” + the modified USER name where non-alphnum chars are mapped to “_”. Thus, “rouser x@y” and “rwuser x%y” both become a member of the group “grpx_y”. Then, the “grpx_y” group has the write access. So, “x@y” user gets the write access.

**Comment by tzhang06** [comment]
I agree with you that putting up a list of restricted characters and a limit on the string length will be a good first step.
